### PR TITLE
Use pidof to make sure the correct pbs_mom process is identified

### PR DESCRIPTION
**Pull request by itkovian**
The problem addressed is the following. Occasionally, we are seeing service pbs_mom restart hanging with momctl unable to connect to a running pbs_mom. We are on EL6/EL7, and the offending code is located in contrib/init.d/pbs_mom.in +36. 

The problem can be reproduced (on a RHEL based system):

~~~~
# service pbs_mom start
Starting TORQUE Mom:                                       [  OK  ]
# echo $$ > /var/spool/pbs/mom_priv/mom.lock 
# pkill pbs_mom
# service pbs_mom stop
Shutting down TORQUE Mom: cannot connect to MOM on node 'localhost', errno=111 (Connection refused)
cannot connect to MOM on node 'localhost', errno=111 (Connection refused)
cannot connect to MOM on node 'localhost', errno=111 (Connection refused)
cannot connect to MOM on node 'localhost', errno=111 (Connection refused)
cannot connect to MOM on node 'localhost', errno=111 (Connection refused)

kill -0 $pid thinks there is a process with the given $pid (true), but it also assumes this is the pbs_mom process (false), and momctl cannot connect to it:

# momctl -d0
cannot connect to MOM on node 'localhost', errno=111 (Connection refused)
~~~~

So basically any sanity check that wants to restart the pbs_mom service is causing it to hang.

### CODE REVIEW

**Comment by itkovian** [comment]
@knielson I am not sure what make CI fail, I've build locally on this commit and things seem to work out. The output on CI stops after 10K lines, but tests seemed to be passing up to that point.

**Comment by knielson** [comment]
What happens with pidof if there is a child process of pbs_mom also running?

File: contrib/init.d/pbs_mom.in [hunk 1]
```diff
@@ -26,26 +28,28 @@ else
 fi
 
 pidof_pbs_mom() {
-   pid="-1"
-   if [ -f $PBS_HOME/mom_priv/mom.lock ];then
-   	pid=`cat $PBS_HOME/mom_priv/mom.lock`
-   fi
-   echo $pid
+	pid=`pidof pbs_mom`
```
Comment (stdweird) [inline contrib/init.d/pbs_mom.in]: this is not safe to use. during a number of `pbs_mom` phases, more then one process is running that will match this query.
and nothing garantees that the pid you get here is the parent pid you are hoping to get.
so we need to store this pid somehow and read it back from a file.
```
[root@node ~]# ps axf |grep pbs_mom
 3331 pts/1    S+     0:00          \_ grep --color=auto pbs_mom
 5784 ?        SLsl   5:13 /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
14904 pts/0    Ss     0:00  \_ /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
14905 pts/0    S      0:00  |   \_ /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
[root@node ~]# pidof pbs_mom
14905 14904 5784
```
and no order from pidof, eg
```
[root@node2402 ~]# service pbs_mom restart
Restarting pbs_mom (via systemctl):                        [  OK  ]
[root@node ~]# ps axf |grep pbs_mom
 3393 pts/1    S+     0:00          \_ grep --color=auto pbs_mom
14904 pts/0    Ss     0:00 /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
14905 pts/0    S      0:00  \_ /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
 3384 ?        SLsl   0:00 /usr/sbin/pbs_mom -H node -p -d /var/spool/pbs
[root@node ~]# pidof pbs_mom
14905 14904 3384
```

**Comment by itkovian** [comment]
This is pretty much back to what we had in the beginning then.

File: contrib/init.d/pbs_mom.in (hunk 1 above)
Comment (stdweird) [inline contrib/init.d/pbs_mom.in]: and now is see @knielson question
well, lots of strange things :smile:

**Comment by stdweird** [comment]
@itkovian well, you can compare if the pid in the lockfile is in the output of the pidof.

**Comment by stdweird** [comment]
@knielson if only pbs_mom had an option to write the parent pid to a file (it's a pity `momctl` is blocking and thus not safe to use)

**Comment by stdweird** [comment]
since only the parent holds an fd to the `mom.lock`, there's path out of this.

**Comment by knielson** [comment]
Stijn,

We can do that.  Make an issue request in github.

Ken

On Wed, Mar 18, 2015 at 9:27 AM, Stijn De Weirdt <notifications@github.com>
wrote:

> @knielson <https://github.com/knielson> if only pbs_mom had an option to
> write the parent pid to a file (it's a pity momctl is blocking and thus
> not safe to use)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/adaptivecomputing/torque/pull/288#issuecomment-83021073>
> .
>



-- 
       [image: Adaptive Computing] <http://www.adaptivecomputing.com>
 [image: Twitter] <http://twitter.com/AdaptiveMoab> [image: LinkedIn]
<http://www.linkedin.com/company/448673?goback=.fcs_GLHD_adaptive+computing_false_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2&trk=ncsrch_hits>
[image:
YouTube] <http://www.youtube.com/adaptivecomputing> [image: GooglePlus]
<https://plus.google.com/u/0/102155039310685515037/posts> [image: Facebook]
<http://www.facebook.com/pages/Adaptive-Computing/314449798572695?fref=ts>
[image:
RSS] <http://www.adaptivecomputing.com/feed>
Ken Nielson Sr. Software Engineer
 +1 801.717.3700 office    +1 801.717.3738 fax
 1712 S. East Bay Blvd, Suite 300     Provo, UT 84606
 www.adaptivecomputing.com

**Comment by knielson** [comment]
Stijn,

Is this pull request ready to be merged?

Ken

**Comment by stdweird** [comment]
@knielson we are running this PR merged with 5.1-dev from begin of april for 3 weeks now, and no issues so far; so i'm in favour of merging it in.

**Comment by knielson** [comment]
3daaa03..022f213  5.0-dev -> 5.0-dev
bb93500..7d20313  5.1-dev -> 5.1-dev
45cbffc..cca5294  develop -> develop
commit 1148f2ce2ca63ad38c366b7a064d42ea414409d9  4.2-dev
